internal/concurrency/concurrency.go

```diff
@@ -0,0 +1,144 @@
+package concurrency
+
+import (
+	"errors"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/speakeasy-api/speakeasy/internal/config"
+	"github.com/speakeasy-api/speakeasy/internal/env"
+)
+
+// ErrLockExists indicates another process already holds the lock
+var ErrLockExists = errors.New("lock already exists: another Speakeasy process is running. You can: 1) wait for it to complete, 2) wait 1 minute for the lock to expire, or 3) manually remove the .lock file in your ~/.speakeasy directory if no other Speakeasy process is running")
+
+// MaxLockAge defines how old a lock can be before it's considered stale (1 minute)
+const MaxLockAge = time.Minute
+
+// AcquireLock attempts to atomically create a lock file
+// Returns nil if successful, ErrLockExists if another process holds the lock
+func AcquireLock() error {
+	if env.IsConcurrencyLockDisabled() {
+		return nil
+	}
+
+	speakeasyHomeDir, err := config.GetSpeakeasyHomeDir()
+	if err != nil {
+		return err
+	}
+
+	lockFile := filepath.Join(speakeasyHomeDir, ".lock")
+
+	// Check if lock exists and is stale
+	if _, err := os.Stat(lockFile); err == nil {
+		// Lock exists, check if it's stale
+		if isStale, _ := isLockStale(lockFile); isStale {
+			// Lock is stale, remove it
+			os.Remove(lockFile)
+		} else {
+			return ErrLockExists
+		}
+	}
+
+	// O_EXCL with O_CREATE ensures the call fails if the file already exists
+	// This makes the check-and-create atomic
+	file, err := os.OpenFile(lockFile, os.O_CREATE|os.O_EXCL|os.O_WRONLY, 0o600)
```

Comment:
Not actually concurrent safe since there is a read and then write better to use system Flock?

Reply:
FYI flock introduces platform-specific code that can cause issues on operating systems and filesystems that don't natively support file locks. In my previous experience I've found it to be more hassle than its worth as customer base and usage expand, unless its critical path. Maybe phrased differently, I like it much better for internal use cases since its easy to control what OS/FS are in use vs externally built and shipped binaries.

Reply:
Isn't that library intending to solve the crossplatform case though? It's specialized for that?